tools/2025-03-workflow-changes.md

```diff
@@ -0,0 +1,260 @@
+```
+
+_Fig. 1: The developer and PR reviewer steps_
+
+## Design
+
+### High Level Workflow Design Principles
+
+#### Use GitHub workflows for CI/CD setup and workflow layout
+
+We will use GitHub workflows for these things:
+
+* Identity and Security
+* Test runner setup
+* Control flow, including matrix operations
+
+GitHub workflows will not contain core logic (unless the workflow is very simple).
+
+#### Core logic is debuggable and testable on a developer machine with minimal setup
+
+```mermaid
+flowchart TD
+    ghsetup["GitHub Workflow Setup"]:::io
+    devsetup["Dev Machine Setup"]:::io
+    subgraph logic["Core Logic"]
+        cloudsetup["Cloud Setup"]
+        work["Do Work"]
+        cloudcleanup["Cloud Cleanup"]
+    end
+    ghcleanup["GitHub Workflow Cleanup"]:::io
+    devcleanup["Dev Machine Cleanup"]:::io
+    
+    ghsetup --> logic
+    devsetup --> logic
+    cloudsetup --> work
+    work --> cloudcleanup
+    logic --> ghcleanup
+    logic --> devcleanup
+
+    classDef io stroke-dasharray: 5 5
+```
+
+_Fig. 2: Testable automation that can run from a GitHub workflow or a dev machine_
+
+#### Workflows can be run on any fork without requiring access to the `radius-project` repos, and without having a Radius GitHub security role.
+
+```mermaid
+flowchart LR
+    fork["Fork the repo"]
+    setupCreds["Set up credentials"]
+    invoke["Invoke the forked workflow"]
+    
+    fork --> setupCreds
+    setupCreds --> invoke
+```
+
+_Fig. 3: Running a GitHub workflow on a fork_
+
+#### Use the GitHub CLI for default security context
+
+On a developer machine, if the dev is logged into the GitHub CLI, the security context is automatically passed to GitHub by the CLI. For any action that is taken in GitHub we should default to using the GitHub CLI so that this context works locally and within workflows.
+
+In some cases we will choose to use a GitHub CLI call instead of a GitHub provided workflow action so that we can have runnable and testable code on a developer machine (because GitHub actions are not locally testable.)
+
+#### Use GitHub actions for setup, cleanup, and matrix operations
+
+The use of actions should be limited to setup and cleanup that is unique to GitHub workflows or runners because any logic that exists in a GitHub action cannot be executed on a developer machine. Some examples are below:
+
+* Cloning repos
+* Installing tools
+* Retrieving stored secrets
+* Publishing results
+
+> NOTE: The above examples are also needed on a developer machine, but the Implementation is different.
+
+:warning: When using GitHub actions that are published on other repositories, we are [placing our trust](https://arstechnica.com/information-technology/2025/03/supply-chain-attack-exposing-credentials-affects-23k-users-of-tj-actions/) in the authors of that repo that they will prevent malicious code from executing. Choose wisely, and consider forking and customizing a GitHub action instead of calling it directly.
```

Comment:
> Choose wisely, and consider forking and customizing a GitHub action instead of calling it directly.

Could we add more specific guidance here?

Reply:
I added a statement about how to think about which actions to trust.